ambry-store/src/main/java/com.github.ambry.store/Log.java

```diff
@@ -13,161 +13,392 @@
    */
-  void writeFrom(ReadableByteChannel channel, long offset, long size)
+  private void checkArgsAndAllocateFirstSegment(long totalCapacity, long segmentCapacity)
       throws IOException {
-    logger.trace("Log : {} currentWriteOffset {} capacityInBytes {} sizeToAppend {} offset to append at {}",
-        file.getAbsolutePath(), currentWriteOffset, capacityInBytes, size, offset);
-    if (offset < 0 || offset + size > currentWriteOffset.get()) {
-      metrics.overflowWriteError.inc(1);
-      throw new IllegalArgumentException("Log : " + file.getAbsolutePath() + " error trying to write to log " +
-          "from channel since new data size " + size + "exceeds log end offset " + currentWriteOffset.get());
+    if (totalCapacity <= 0 || segmentCapacity <= 0) {
+      throw new IllegalArgumentException(
+          "One of totalCapacityInBytes [" + totalCapacity + "] or " + "segmentCapacityInBytes [" + segmentCapacity + "]"
+              + " is <=0");
     }
-    long bytesWritten = 0;
-    while (bytesWritten < size) {
-      bytesWritten += fileChannel.transferFrom(channel, offset + bytesWritten, size - bytesWritten);
+    segmentCapacity = Math.min(totalCapacity, segmentCapacity);
+    // all segments should be the same size.
+    long numSegments = totalCapacity / segmentCapacity;
+    if (totalCapacity % segmentCapacity != 0) {
+      throw new IllegalArgumentException(
+          "Capacity of log [" + totalCapacity + "] should be a multiple of segment capacity [" + segmentCapacity + "]");
     }
-    logger.trace("Log : {} bytes written to the log from read channel at {}, bytesWritten: {}", file.getAbsolutePath(),
-        offset, bytesWritten);
+    remainingUnallocatedSegments = numSegments;
+    String firstSegmentName = LogSegmentNameHelper.generateFirstSegmentName(numSegments);
+    File segmentFile = allocate(LogSegmentNameHelper.nameToFilename(firstSegmentName), segmentCapacity);
+    // to be backwards compatible, headers are not written for a log segment if it is the only log segment.
+    LogSegment segment = new LogSegment(firstSegmentName, segmentFile, segmentCapacity, metrics, numSegments > 1);
+    segmentsByName.put(segment.getName(), segment);
   }
 
   /**
-   * Close this log
+   * Loads segment files and creates {@link LogSegment} instances.
+   * @param segmentFiles the files that form the segments of the log.
+   * @param totalCapacity the total capacity of the log. This is used only if this is a single segment log.
+   * @throws IOException if there is an I/O error loading the segment files or creating {@link LogSegment} instances.
    */
-  void close()
+  private void loadSegments(File[] segmentFiles, long totalCapacity)
+      throws IOException {
+    long totalSegments = -1;
+    for (File segmentFile : segmentFiles) {
+      String name = LogSegmentNameHelper.nameFromFilename(segmentFile.getName());
+      LogSegment segment;
+      if (name.isEmpty()) {
+        totalSegments = 1;
```

Comment:
It cannot be created - I don't want to give the impression that it can be. I don't think it is possible to check for everything i.e. why would there be other files with "log_current" if there is no code creating it? Would prefer not to add those checks. Let me know if you feel strongly about this.